### PR TITLE
[ci] prevent `lgb.model` and `lgb.pkl` files being left behind after testing

### PR DESCRIPTION
**Pull request by nicklamiller**
contributes to: #6361

Specificially prevents `lgb.pkl` and `lgb.model` files from being left behind after running tests by using pytest's `tmp_path` fixture. One can use the reprex described in https://github.com/microsoft/LightGBM/issues/6361#issue-2187570183 to confirm these files are no longer left behind after these changes.

### CODE REVIEW

File: tests/python_package_test/test_engine.py [hunk 1]
```diff
@@ -70,6 +70,11 @@ def categorize(continuous_x):
     return np.digitize(continuous_x, bins=np.arange(0, 1, 0.01))
 
 
+@pytest.fixture
+def model_path_txt(tmp_path):
+    return str(tmp_path / "lgb.model")
```
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:75]: Can you please revert this? I don't think the extra indirection of a separate fixture gets us much here.

`pytest's` `tmp_path` fixture will resolve to a different path for every test case, so we're well-protected already against two tests writing to / reading from the same paths. Please just use `tmp_path` directly in the test cases.

File: tests/python_package_test/test_engine.py [hunk 2]
```diff
@@ -1258,14 +1263,13 @@ def test_cv():
...
 
-def test_cv_works_with_init_model(tmp_path):
+def test_cv_works_with_init_model(model_path_txt):
     X, y = make_synthetic_regression()
     params = {"objective": "regression", "verbose": -1}
     num_train_rounds = 2
     lgb_train = lgb.Dataset(X, y, free_raw_data=False)
     bst = lgb.train(params=params, train_set=lgb_train, num_boost_round=num_train_rounds)
     preds_raw = bst.predict(X, raw_score=True)
-    model_path_txt = str(tmp_path / "lgb.model")
```
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:1268]: Since this was already using `tmp_path`, I think it's very unlikely that this one was leaving a file behind. Could you please revert the changes to this case and any others that were already using `tmp_path`?

File: tests/python_package_test/test_engine.py [hunk 3]
```diff
@@ -1432,7 +1434,7 @@ def test_feature_name():
     assert feature_names == gbm.feature_name()
 
 
-def test_feature_name_with_non_ascii(rng):
+def test_feature_name_with_non_ascii(rng, model_path_txt):
```
Comment (jameslamb) [inline tests/python_package_test/test_engine.py:1437]: As I mentioned above... let's please just use `tmp_path` here.

Assign a variable with the model file path and re-use it in both `gbm.save_model()` and `lgb.Booster()` below, something like this:

```python
model_path_txt = tmp_path / "lgb.model"
gbm.save_model(model_path_txt)

gbm2 = lgb.Booster(model_file=model_path_txt)
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
Thanks so much for working on this! Please see my first round of suggestions.

File: tests/python_package_test/test_engine.py (hunk 1 above)
Comment (nicklamiller) [reply tests/python_package_test/test_engine.py]: Yeah I suppose the fixture was a bit overkill, reverted in [50f4893](https://github.com/microsoft/LightGBM/pull/6518/commits/50f48934695d82707dff6c58b643ef59a0e0128f) and I believe this addresses the other suggestions as well.

Comment (jameslamb) [reply tests/python_package_test/test_engine.py]: No problem! This one is definitely a matter of personal preference. Thanks for reverting that.

**Review by jameslamb — APPROVED** [review]
Thanks very much! I pulled this branch and ran the tests, confirmed that there now `lgb.model` and `lgb.pkl` aren't left behind.
